File: components/devtools_bridge/android/java/src/org/chromium/components/devtools_bridge/apiary/ApiaryClientFactory.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.devtools_bridge.apiary;

import android.net.http.AndroidHttpClient;

import org.chromium.base.JNINamespace;

/**
 * Factory for creating clients for external APIs.
 */
@JNINamespace("devtools_bridge::android")
public class ApiaryClientFactory {
    private static final String USER_AGENT = "DevTools bridge";

    public static final String OAUTH_SCOPE = "https://www.googleapis.com/auth/clouddevices";

    protected final AndroidHttpClient mHttpClient = AndroidHttpClient.newInstance(USER_AGENT);

    /**
     * Creates a new GCD client with auth token.
     */
    public GCDClient newGCDClient(String oAuthToken) {
        return new GCDClient(mHttpClient, getAPIKey(), oAuthToken);
    }

    /**
     * Creates a new anonymous client. GCD requires client been not authenticated by user or
     * device credentials for finalizing registration.
     */
    public GCDClient newAnonymousGCDClient() {
        return new GCDClient(mHttpClient, nativeGetAPIKey());
    }

    public OAuthClient newOAuthClient() {
        return new OAuthClient(
                mHttpClient, OAUTH_SCOPE, getOAuthClientId(), nativeGetOAuthClientSecret());
    }

    public BlockingGCMRegistrar newGCMRegistrar() {
        return new BlockingGCMRegistrar();
    }

    public void close() {
        mHttpClient.close();
    }

    public String getOAuthClientId() {
        return nativeGetOAuthClientId();
    }

    protected String getAPIKey() {
        return nativeGetAPIKey();
    }

    private native String nativeGetAPIKey();
    private native String nativeGetOAuthClientId();
    private native String nativeGetOAuthClientSecret();
}

File: components/devtools_bridge/android/apiary_client_factory.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/devtools_bridge/android/apiary_client_factory.h"

#include "base/android/jni_string.h"
#include "google_apis/google_api_keys.h"
#include "jni/ApiaryClientFactory_jni.h"

using base::android::ConvertUTF8ToJavaString;

namespace devtools_bridge {
namespace android {

// static
bool ApiaryClientFactory::RegisterNatives(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

// static
jstring GetAPIKey(JNIEnv* env, jobject jcaller) {
  return ConvertUTF8ToJavaString(env, google_apis::GetAPIKey()).Release();
}

// static
jstring GetOAuthClientId(JNIEnv* env, jobject jcaller) {
  return ConvertUTF8ToJavaString(
             env, google_apis::GetOAuth2ClientID(
                      google_apis::OAuth2Client::CLIENT_MAIN)).Release();
}

// static
jstring GetOAuthClientSecret(JNIEnv* env, jobject jcaller) {
  return ConvertUTF8ToJavaString(
             env, google_apis::GetOAuth2ClientSecret(
                      google_apis::OAuth2Client::CLIENT_MAIN)).Release();
}

}  // namespace android
}  // namespace devtools_bridge
